File: src/lib/ingest/readme.ts
/**
 * README discovery in the root of the default-branch tree.
 */
import type { RawTreeEntry } from './tree';

/** Rank for README candidates: lower is better; null ⇒ not a readme. */
export function readmeRank(name: string): number | null {
  const l = name.toLowerCase();
  if (l === 'readme.md') return 0;
  if (l === 'readme.markdown' || l === 'readme.mdown') return 1;
  if (l === 'readme') return 2;
  if (l === 'readme.txt') return 3;
  if (l === 'readme.rst' || l === 'readme.adoc' || l === 'readme.org') return 4;
  if (l.startsWith('readme.')) return 5;
  return null;
}

/** Pick the README entry among root entries (prefers README.md). */
export function findReadmeEntry(rootEntries: RawTreeEntry[]): RawTreeEntry | null {
  const candidates = rootEntries
    .filter((e) => e.type === 'blob' && readmeRank(e.name) !== null)
    .sort((a, b) => readmeRank(a.name)! - readmeRank(b.name)! || (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  return candidates[0] ?? null;
}

File: internal/ingest/readme.go
package ingest

import (
	"sort"
	"strings"
)

// readmeRank ranks a README candidate found in the root of the default-branch tree; lower is
// better, and -1 means "not a readme at all".
//
// The port of readmeRank in src/lib/ingest/readme.ts, where the function returns `number |
// null`. -1 stands in for null because every caller only ever asks "is this a readme" and
// "which of these two is better", and a sentinel keeps both answers in one comparison.
func readmeRank(name string) int {
	l := strings.ToLower(name)
	switch {
	case l == "readme.md":
		return 0
	case l == "readme.markdown" || l == "readme.mdown":
		return 1
	case l == "readme":
		return 2
	case l == "readme.txt":
		return 3
	case l == "readme.rst" || l == "readme.adoc" || l == "readme.org":
		return 4
	case strings.HasPrefix(l, "readme."):
		return 5
	}
	return -1
}

// findReadmeEntry picks the README among root entries, preferring README.md.
//
// Ties in rank fall back to the name so the pick never depends on the order git listed the
// tree in — ls-tree is already path-sorted, but a rank tie between README.rst and README.adoc
// would otherwise be settled by whatever came first.
func findReadmeEntry(rootEntries []RawTreeEntry) *RawTreeEntry {
	candidates := []RawTreeEntry{}
	for _, e := range rootEntries {
		if e.Type == "blob" && readmeRank(e.Name) >= 0 {
			candidates = append(candidates, e)
		}
	}
	if len(candidates) == 0 {
		return nil
	}
	sort.SliceStable(candidates, func(i, j int) bool {
		ri, rj := readmeRank(candidates[i].Name), readmeRank(candidates[j].Name)
		if ri != rj {
			return ri < rj
		}
		return candidates[i].Name < candidates[j].Name
	})
	best := candidates[0]
	return &best
}
